### PR TITLE
[SYCLomatic] Use UnifiedPath to unify the canonicalization of path

### PR DESCRIPTION
**Pull request by zhiweij1**
Signed-off-by: Jiang, Zhiwei <zhiwei.jiang@intel.com>

### CODE REVIEW

File: clang/include/clang/Tooling/Tooling.h [hunk 1]
```diff
@@ -196,8 +194,8 @@ inline std::unique_ptr<FrontendActionFactory> newFrontendActionFactory(
 ///                         clang modules.
 ///
 /// \return - True if 'ToolAction' was successfully executed.
-bool runToolOnCode(std::unique_ptr<FrontendAction> ToolAction, const Twine &Code,
-                   const Twine &FileName = "input.cc",
+bool runToolOnCode(std::unique_ptr<FrontendAction> ToolAction,
+                   const Twine &Code, const Twine &FileName = "input.cc",
```
Comment (tomflinda) [inline clang/include/clang/Tooling/Tooling.h]: Public code should not be touched.

**Review by tomflinda — APPROVED** [review]
LGTM

File: clang/include/clang/Tooling/Tooling.h [hunk 2]
```diff
@@ -35,6 +35,7 @@
 #include "clang/Frontend/FrontendAction.h"
 #include "clang/Frontend/PCHContainerOperations.h"
 #include "clang/Tooling/ArgumentsAdjusters.h"
+#include "clang/Tooling/Core/DpctPath.h"
```
Comment (zhiweij1) [inline clang/include/clang/Tooling/Tooling.h]: SYCLomatic_CUSTOMIZATION

File: clang/lib/DPCT/AnalysisInfo.cpp [hunk 3]
```diff
@@ -2857,8 +2848,14 @@ void CallFunctionExpr::buildInfo() {
   //
   // TODO: Need to revisit the condition to add SYCL_EXTERNAL macro if issues
   // are observed in the future.
-  if (!DefFilePath.empty() && DefFilePath != getFilePath() &&
+  std::cout << "!!!!!!!!!!!!! FuncInfo.get():" << FuncInfo.get() << std::endl;
```
Comment (zhiweij1) [inline clang/lib/DPCT/AnalysisInfo.cpp]: remove it

File: clang/lib/DPCT/AnalysisInfo.cpp [hunk 4]
```diff
@@ -3217,10 +3215,14 @@ DeviceFunctionDecl::DeviceFunctionDecl(unsigned Offset,
     FuncInfo = std::make_shared<DeviceFunctionInfo>(
         FD->param_size(), NonDefaultParamNum, getFunctionName(FD));
   }
-  if (!FilePath.empty()) {
+  std::cout << "!!!!!!!!!!!!!!! DeviceFunctionDecl FilePath.getCanonicalPath().str():" << FilePath.getCanonicalPath().str() << std::endl;
```
Comment (zhiweij1) [inline clang/lib/DPCT/AnalysisInfo.cpp]: remove print

File: clang/lib/DPCT/AnalysisInfo.h [hunk 5]
```diff
@@ -637,11 +638,13 @@ class DpctFileInfo {
   std::unordered_set<std::shared_ptr<TextModification>> ConstantMacroTMSet;
   std::unordered_map<std::string, std::tuple<unsigned int, std::string, bool>>
       AtomicMap;
+public:
```
Comment (zhiweij1) [inline clang/lib/DPCT/AnalysisInfo.h]: revert this change

File: clang/include/clang/Tooling/Tooling.h (hunk 2 above)
Comment (zhiweij1) [reply clang/include/clang/Tooling/Tooling.h]: done

File: clang/lib/DPCT/AnalysisInfo.h (hunk 5 above)
Comment (zhiweij1) [reply clang/lib/DPCT/AnalysisInfo.h]: done

File: clang/include/clang/Tooling/Tooling.h (hunk 1 above)
Comment (zhiweij1) [reply clang/include/clang/Tooling/Tooling.h]: done

File: clang/lib/Tooling/Core/DpctPath.cpp [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+    llvm::sys::path::append(TempPath, llvm::sys::path::Style::native, Path);
+    Path = TempPath;
+  }
+
+  llvm::sys::path::remove_dots(Path, /* remove_dot_dot= */ true);
+  llvm::sys::path::native(Path);
+
+  llvm::SmallString<512> RealPath;
+  // We need make sure the input `Path` for llvm::sys::fs::real_path is
+  // exsiting.
```
Comment (zhimingwang36) [inline clang/lib/Tooling/Core/DpctPath.cpp]:
```suggestion
  // exsiting, or else the behaviour of real_path() is unexpected.
```

File: clang/include/clang/Tooling/Core/DpctPath.h [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+#include "llvm/ADT/Twine.h"
+#include "llvm/Support/FileSystem.h"
+#include "llvm/Support/Path.h"
+
+#include <string>
+
+namespace clang {
+namespace tooling {
+
+class DpctPath {
```
Comment (zhimingwang36) [inline clang/include/clang/Tooling/Core/DpctPath.h]:
```suggestion
class UnifiedPath {
```

File: clang/lib/DPCT/Utility.h [hunk 8]
```diff
@@ -112,131 +113,94 @@ template <class StreamT> class CurlyBracketsPrinter {
   ~CurlyBracketsPrinter() { OS << "}"; }
 };
 
-bool makeCanonical(llvm::SmallVectorImpl<char> &Path);
-bool makeCanonical(std::string &Path);
-bool isCanonical(llvm::StringRef Path);
+//bool makeCanonical(llvm::SmallVectorImpl<char> &Path);
+//bool makeCanonical(std::string &Path);
+//bool isCanonical(llvm::StringRef Path);
```
Comment (zhimingwang36) [inline clang/lib/DPCT/Utility.h]: please remove it if not used anymore.

File: clang/include/clang/Tooling/Core/DpctPath.h (hunk 7 above)
Comment (zhiweij1) [reply clang/include/clang/Tooling/Core/DpctPath.h]: done

File: clang/lib/DPCT/Utility.h (hunk 8 above)
Comment (zhiweij1) [reply clang/lib/DPCT/Utility.h]: done

File: clang/lib/Tooling/Core/DpctPath.cpp (hunk 6 above)
Comment (zhiweij1) [reply clang/lib/Tooling/Core/DpctPath.cpp]: done

File: clang/lib/DPCT/AnalysisInfo.cpp (hunk 4 above)
Comment (zhiweij1) [reply clang/lib/DPCT/AnalysisInfo.cpp]: done

File: clang/lib/DPCT/AnalysisInfo.cpp (hunk 3 above)
Comment (zhiweij1) [reply clang/lib/DPCT/AnalysisInfo.cpp]: done
